### PR TITLE
refactor: move logic out of main `PaperListingForm` file

### PR DESCRIPTION
**Pull request by jaredcwhite**
## Issue

Addresses #1916 

- [ ] This change addresses the issue in full
- [x] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

(there are more suggestions in the issue I'd love to pursue, but this is a solid first go-round.)

## Description

In the interests of breaking up the large `sites/partners/src/listings/PaperListingForm/index.tsx` file, this PR contains the following changes:

* Default form data and types/interfaces have been relocated to `formTypes.ts`
* The previous `formatFormData` has been refactored out into the `ListingDataPipeline` class, which in turns processes form data with a list of formatters.
* The `Formatter` class and all subclassed formatters now live in the `src/listings/PaperListingForm/formatters` folder.
* New formatters can be created and added to the pipeline easily.
* A new Jest unit test suite has been added `yarn test:unit` and a significant portion of the formatters now have accompanying unit tests.
* The pipeline/formatter architecture is listing form-agnostic, so we can easily port it to other large forms in the future if necessary.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [x] This change is a refactor/addresses technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Besides running unit tests, just saving new/existing listings in the Listing Form.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have updated the changelog to include a description of my changes
- [ ] I have run `yarn generate:client` if I made backend changes
- [ ] I have exported any new pieces in ui-components

### CODE REVIEW

**Comment by emilyjablonski** [comment]
This code is much cleaner and separated, thank you! I think overall I am finding the move to the class with individual formatters to be more confusing to read than how it was before though. Can you outline what you think the benefits are of this being the class/formatter type?

**Review by dominikx96 — CHANGES_REQUESTED** [review]
@jaredcwhite I noticed a problem with open houses, I've filled out the whole form, but values (date, start, end) don't exist - on dev it works as expected, see:

Partners Portal 2021-11-10 at 2 05 23 PM

File: sites/partners/src/listings/PaperListingForm/formatters/BooleansFormatter.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import Formatter from "./Formatter"
```
[the remark this answers is not in the document]
Comment (jaredcwhite) [reply sites/partners/src/listings/PaperListingForm/formatters/BooleansFormatter.ts:1]: I was hoping to find a way to cut down the repetition here, but I wasn't able to find a fool-proof solution. Before it was a just a lot of nested ternary operators which is pretty gnarly, so this is at least a little more obvious what's happening. Happy for any suggestions on how to improve…

**Comment by jaredcwhite** [comment]
> This code is much cleaner and separated, thank you! I think overall I am finding the move to the class with individual formatters to be more confusing to read than how it was before though. Can you outline what you think the benefits are of this being the class/formatter type?

@emilyjablonski Sure, so the first win would be that adding new formatters is trivial — just add a new file in `formatters` that subclasses `Formatter` and modifies `this.data` in some way, add a single new `import` to `ListingDataPipeline`, and the rest falls into place automatically. In addition, any shared logic (like `processBoolean`) can be added to `Formatter` directly and then all the subclasses would inherit that logic.

The second win is because we have standalone formatters, they can all be unit tested independently of each other. I didn't completely fill out the tests yet in this PR, but it's a decent part of the way done.

The third, downstream win is we can apply a similar pattern anywhere else in the frontend codebase where such a data pipeline is required.

If there are any particular areas which are hard to parse in the new structure, feel free to suggest what you'd like to see improved.

**Comment by emilyjablonski** [comment]
@jaredcwhite Do you mind checking out Dominik's comment above?

**Comment by emilyjablonski** [comment]
@jaredcwhite Another bump :)

**Comment by jaredcwhite** [comment]
Yep, I can work on refreshing this now that we have a bunch of new form stuff merged into dev.

File: sites/partners/src/listings/PaperListingForm/formatters/BooleansFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    this.processBoolean("applicationPickUpAddressType", {
+      when:
+        this.data.canPaperApplicationsBePickedUp === YesNoAnswer.Yes &&
+        addressTypes[this.data.whereApplicationsPickedUp] !== addressTypes.anotherAddress,
+      yes: () => addressTypes[this.data.whereApplicationsPickedUp],
+    })
+    this.processBoolean("applicationDropOffAddress", {
+      when:
+        this.data.canApplicationsBeDroppedOff === YesNoAnswer.Yes &&
+        this.data.whereApplicationsPickedUp === addressTypes.anotherAddress,
```
[the remark this answers is not in the document]
Comment (jaredcwhite) [reply sites/partners/src/listings/PaperListingForm/formatters/BooleansFormatter.ts]: Yes, good catch!

**Comment by jaredcwhite** [comment]
> @jaredcwhite I noticed a problem with open houses, I've filled out the whole form, but values (date, start, end) don't exist - on dev it works as expected, see:
> 
> Partners Portal 2021-11-10 at 2 05 23 PM

@dominikx96 I'm seeing an issue with date parsing in general now, just for the open houses, and it's on dev too! 😬 Look here:

Screen Shot 2021-12-01 at 8 13 08 PM

**Comment by jaredcwhite** [comment]
Welp, there's another issue with open house events that's on dev! #2265

File: sites/partners/src/listings/PaperListingForm/formatters/BooleansFormatter.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (seanmalbert) [reply sites/partners/src/listings/PaperListingForm/formatters/BooleansFormatter.ts:1]: I think I'm good with this too, any abstraction to reduce repetition would probably make it more difficult to modify or at least read.

**Review by seanmalbert — APPROVED** [review]
This looks good, @jaredcwhite! I think there's an issue with units not persisting in the case of a new listing and field validation failing on publish, but I don't think it's related to this.
